### PR TITLE
Split the various modules of the build script, to permit a customizat…

### PR DESCRIPTION
**Pull request by guruDanny67**
…ions of the projects/tools to build

Fist tentative of the split. 

This should be exactly compatible and equal to the old, single, build.py script. The only difference are a check if we try to add a project with the same name of one already registered and the explanation of the NotImplemented exception in Project.build().

### CODE REVIEW

File: build/build_utils.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+class Tarball(object):
+    def unpack(self):
+        print_log('Extracting %s to %s' % (self.archive_file, self.builder.working_dir))
+
+        os.makedirs(self.build_dir)
+        self.builder.exec_msys([self.builder.tar, 'ixf', convert_to_msys(self.archive_file), '-C', convert_to_msys(self.build_dir), '' if self.tarbomb else '--strip-components=1'])
+
+        print_log('Extracted %s' % (self.archive_file,))
+
+class MercurialRepo(object):
```
Comment (nacho) [inline build/build_utils.py]: I think we could split the repos as well

File: build/build_utils.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+            self.builder.exec_msys('git pull --rebase', working_dir=self.build_dir)
+
+        if self.fetch_submodules:
+            self.builder.exec_msys('git submodule update --init', working_dir=self.build_dir)
+
+        if os.path.exists(self.patch_dir):
+            print_log("Copying files from %s to %s" % (self.patch_dir, self.build_dir))
+            self.builder.copy_all(self.patch_dir, self.build_dir)
+
+class Project(object):
```
Comment (nacho) [inline build/build_utils.py]: same for project, this could be added to a project file

File: build/build_utils.py (hunk 1 above)
Comment (guruDanny67) [inline build/build_utils.py]: Like having a hg_repo.py and git_repo.py or a single file, say build_unpack.py, with tarball, hg, git (and, maybe in the future, zipball or whatever) ?

File: build/build_utils.py (hunk 2 above)
Comment (guruDanny67) [inline build/build_utils.py]: Single file with the Project class, right (and also the Tool one, these are the base building blocks ...) ? 

The Meson and CmakeProject where should go, in a base_builders.py module ?

Thanks

**Comment by guruDanny67** [comment]
Done another pull request, sincronized with the new use of the internal Tar python module.

**Pull request closed without merging**